### DISCUSSION TITLE
Implementing unsupervised Domain Adversarial Training in SpeechBrain

### DISCUSSION THREAD

**Discussion by GryffindorLi** [speechbrain/speechbrain#1752] [Q&A]
Hi,

I'm trying to implement an unsupervised Domain Adversarial Training for accented speech in SpeechBrain. 

I would like to pass 2 dataloaders to the `Brain` class. One for labeled source domain data and the other for unlabeled target domain data. But I have no idea how to implement this. Can anyone help me with this?

**Comment by anautsch** [comment]
Hi @GryffindorLi good question.

There might be several ways to tackle this. Here's one. Please take a look at
https://github.com/speechbrain/speechbrain/blob/5df3885c5f2f908433757a21e25a3bd4732c57ec/speechbrain/dataio/dataset.py#L383

which needs to set the following to be functional
```python
    def __init__(self, from_dataset, data_ids):
        self.data = from_dataset.data
        self.data_ids = data_ids
        self.pipeline = copy.deepcopy(from_dataset.pipeline)
```
so, it's just about getting in both of the un/labeled data; their IDs and to put output keys wheter/not they are un/labeled. 

Then, you can either use such a sampling as-is to randomly mix un/labeled data (somewhat random; there will be different proportions across your datasets), -or- you might treat them differently during each batch. During each batch, you can extract from its signature whether/not the data is un/labeled as of provided output keys & you'll know the training stage, also.

It's just an idea. You might want to start with a minimal unittest to verify your solution before running a full recipe (it saves your time to get the sampling done right first). This integration test is on sorting when DDP is used, but it might be a good starting point for you to scrap out what you don't need and to continue from for what you need.

If this helps, please seek into providing a recipe PR, so the community has the solution available as well.

**Reply by GryffindorLi** [reply]
Thank you for your reply! I will looking to your suggestion!
